### PR TITLE
nice: move from getopts to clap #1794

### PR DESCRIPTION
**Pull request by jeckersb**

### CODE REVIEW

File: src/uu/nice/src/nice.rs [hunk 1]
```diff
@@ -26,64 +26,49 @@ extern "C" {
...
-    }
+pub fn uumain(args: impl uucore::Args) -> i32 {
+    let usage = get_usage();
+
+    let matches = App::new(executable!())
+        .setting(AppSettings::TrailingVarArg)
+        .version(VERSION)
+        .usage(&usage[..])
+        .arg(
+            Arg::with_name("adjustment")
```
Comment (sylvestre) [inline src/uu/nice/src/nice.rs:50]: For consistency with other binaries, please declare adjustment in a variable (or structure)

File: src/uu/nice/src/nice.rs [hunk 2]
```diff
@@ -26,64 +26,49 @@ extern "C" {
...
+                .short("n")
+                .long("adjustment")
+                .help("add N to the niceness (default is 10)")
+                .takes_value(true)
+                .allow_hyphen_values(true),
+        )
+        .arg(Arg::with_name("COMMAND").multiple(true))
```
Comment (sylvestre) [inline src/uu/nice/src/nice.rs:57]: ditto

File: src/uu/nice/src/nice.rs [hunk 3]
```diff
@@ -26,64 +26,49 @@ extern "C" {
...
+        .get_matches_from(args);
 
     let mut niceness = unsafe { getpriority(PRIO_PROCESS, 0) };
     if Error::last_os_error().raw_os_error().unwrap() != 0 {
         show_error!("{}", Error::last_os_error());
         return 125;
     }
 
-    let adjustment = match matches.opt_str("adjustment") {
+    let adjustment = match matches.value_of("adjustment") {
```
Comment (sylvestre) [inline src/uu/nice/src/nice.rs:66]: and update this :)

File: tests/by-util/test_nice.rs [hunk 4]
```diff
@@ -1 +1,70 @@
...
+}
+
+#[test]
+fn test_negative_adjustment() {
+    // This assumes the test suite is run as a normal (non-root) user, and as
+    // such attempting to set a negative niceness value will be rejected by
+    // the OS.  If it gets denied, then we know a negative value was parsed
+    // correctly.
+
+    let res = new_ucmd!().args(&["-n", "-1", "true"]).run();
```
Comment (sylvestre) [inline tests/by-util/test_nice.rs:31]: thanks :)

**Review by sylvestre — COMMENTED** [review]
Well done, just some small nits !

File: src/uu/nice/src/nice.rs [hunk 5]
```diff
@@ -26,64 +26,54 @@ extern "C" {
...
-
-    if matches.opt_present("version") {
-        println!("{} {}", NAME, VERSION);
-        return 0;
-    }
-
-    if matches.opt_present("help") {
-        let msg = format!(
-            "{0} {1}
+pub mod options {
```
Comment (sylvestre) [inline src/uu/nice/src/nice.rs:29]: Merci!

**Comment by sylvestre** [comment]
This is failing on mac os, could you please fix this? thanks

**Comment by jeckersb** [comment]
> This is failing on mac os, could you please fix this? thanks

Yeah I was just looking into that.  I don't have a macos system handy that I can test on.  But as far as I can tell from https://opensource.apple.com/source/shell_cmds/shell_cmds-17.1/nice/nice.1.auto.html, macos is using the netbsd version of nice.  So https://man.netbsd.org/nice.1 if you want something human-readable.  The main issue seems to be that the BSD version of nice doesn't seem to support running a bare "nice" without any arguments to display the current level of niceness.  So a lot of the tests are failing due to that.  But I think there's additional issues too beyond that.

I'll have to spin up a BSD VM to play around and that should be enough to approximate the behavior on macos, then I can sort out the differences there.

**Comment by sylvestre** [comment]
You can always push changes in your github fork to debug remotely (not ideal :)

**Review by sylvestre — CHANGES_REQUESTED** [review]
mac failures

**Comment by jeckersb** [comment]
Few updates here:

- I ditched the few tests that were doing some variation of "nice nice" because the 2nd invocation is dependent on the system version, and that's not really what we want to test anyway.
- I spun up a netbsd virtual machine, and the tests pass there.  Yay.
- I got my hands on a macbook, and... the tests also pass on the macbook???

At this point I am quite confused why the macos CI is failing.

**Comment by jeckersb** [comment]
Finally figured this one out.  `nice` was checking `errno` after calling `getpriority`, but wasn't clearing `errno` beforehand.  When run inside of the test suite, some previous library call in another utility must have been setting `errno` and then the `getpriority` call was mistakenly being interpreted as failed.  No matter what arguments were supplied to `nice`, it would always fail with `ENOENT` (which isn't even a valid error for getpriority).  If `nice` was run standalone, it worked.  It also worked if the tests were invoked like `cargo test --features nice --no-default-features` because there was no prior library call to set `errno`.  The fact that this worked on platforms other than macos was just luck.

With that said, @sylvestre I think this is ready now.

**Review by sylvestre — APPROVED** [review]
well done!

**Pull request merged**